protocol/contracts/amo/TpfAmo.sol

```diff
@@ -0,0 +1,549 @@
+     * pool exit. TEMPLE tokens returned from balancer pool are burned
+     * @param bptAmountIn amount of BPT tokens going in balancer pool for exit
+     * @param minAmountOut amount of TEMPLE tokens expected out of balancer pool
+     */
+    function rebalanceUp(
+        uint256 bptAmountIn,
+        uint256 minAmountOut
+    ) external onlyOperatorOrOwner whenNotPaused enoughCooldown {
+        _validateParams(minAmountOut, bptAmountIn);
+
+        // check spot price is below TPF by lower bound
+        if (!poolHelper.isSpotPriceBelowTPFLowerBound()) {
+            revert AMOCommon.NoRebalanceUp();
+        }
+
+        // will exit take price above tpf + upper bound
+        // should rarely be the case, but a sanity check nonetheless
+        if (poolHelper.willExitTakePriceAboveTPFUpperBound(minAmountOut)) {
+            revert AMOCommon.HighSlippage();
+        }
+
+        // construct request
+        AMO__IBalancerVault.ExitPoolRequest memory exitPoolRequest;
+        if (templeBalancerPoolIndex == 0) {
+            exitPoolRequest = poolHelper.createPoolExitRequest(address(temple), address(stable), bptAmountIn,
+                0, minAmountOut, templeBalancerPoolIndex);
+        } else {
+            exitPoolRequest = poolHelper.createPoolExitRequest(address(temple), address(stable), bptAmountIn,
+                1, minAmountOut, templeBalancerPoolIndex);
+        }
+        // withdraw bpt tokens
+        amoStaking.withdrawAndUnwrap(bptAmountIn, false, true);
+
+        bptToken.approve(address(balancerVault), 0);
+        bptToken.approve(address(balancerVault), bptAmountIn);
+
+        // execute call and check for sanity
+        uint256 templeBalanceBefore = temple.balanceOf(address(this));
+        uint256 spotPriceScaledBefore = poolHelper.getSpotPriceScaled();
+        balancerVault.exitPool(balancerPoolId, address(this), address(this), exitPoolRequest);
+        uint256 templeBalanceAfter = temple.balanceOf(address(this));
+        
+        // burn
+        uint256 burnAmount;
+        unchecked {
+            burnAmount = templeBalanceAfter - templeBalanceBefore;
+        }
+        AMO__ITempleERC20Token(address(temple)).burn(burnAmount);
+
+        if (uint64(poolHelper.getSlippage(spotPriceScaledBefore)) > postRebalanceSlippage) {
+            revert AMOCommon.HighSlippage();
+        }
+        lastRebalanceTimeSecs = uint64(block.timestamp);
+        emit RebalanceUp(bptAmountIn, burnAmount);
+    }
+
+     /**
+     * @notice Rebalance when $TEMPLE spot price is above Treasury Price Floor
+     * Mints TEMPLE tokens and single-side deposits into balancer pool
+     * Returned BPT tokens are deposited and staked into Aura for rewards using the staking contract.
+     * @param templeAmountIn Amount of TEMPLE tokens to deposit into balancer pool
+     * @param minBptOut Minimum amount of BPT tokens expected to receive
+     * 
+     */
+    function rebalanceDown(
```

Comment:
Meta-question - keepers are generally permisionless to call performUpkeep(). How do you envisage working out and passing through the min amounts (off chain to avoid sandwhich) in a way that the keeper can still be called permisionlessly?